### PR TITLE
+str Add Flow#preMaterialize.

### PR DESCRIPTION
**Pull request by He-Pin**
refs #30074
Seems add much overhead

### CODE REVIEW

File: .scalafmt.conf [hunk 1]
```diff
@@ -5,6 +5,7 @@ style = defaultWithAlign
 docstrings                 = JavaDoc
 indentOperator             = spray
 maxColumn                  = 120
+lineEndings                = preserve
```
Comment (johanandren) [inline .scalafmt.conf:8]: Please don't change the formatting rules

**Review by johanandren — COMMENTED** [review]
Somewhat costly indeed (and also has some surprises with subscription timeouts) but that's not different from the other preMaterialize methods. Needs at least some happy path test coverage as well.

File: .scalafmt.conf (hunk 1 above)
Comment (hepin1989) [reply .scalafmt.conf:8]: But I am using Windows, when compile, many file changes.
With this just `keep` the same.

Comment (hepin1989) [reply .scalafmt.conf:8]: Many files is currnetly end with `CRLF`

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala [hunk 2]
```diff
@@ -145,6 +145,16 @@ final class Flow[-In, +Out, +Mat](
...
     new Flow(traversalBuilder.transformMat(f), shape)
 
+  /**
+   * Materializes this [[Flow]], immediately returning (1) its materialized value, and (2) a newly materialized [[Flow]].
+   * The returned flow is partial materialized and do not support multiple times materialization.
+   */
+  def preMaterialize()(implicit materializer: Materializer): (Mat, ReprMat[Out, NotUsed]) = {
+    val ((sub, mat), pub) =
+      Source.asSubscriber[In].viaMat(this)(Keep.both).toMat(Sink.asPublisher(false))(Keep.both).run()
+    (mat, Flow.fromSinkAndSource(Sink.fromSubscriber(sub), Source.fromPublisher(pub)))
```
Comment (hepin1989) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:155]: this can not be the `fromSinkAndSourceCoupled`

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:155]: I think a user wouold expect it to be coupled though, if not, a failure either upstream or downstream will not propagate to the other side once the stream is fully materialized.

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:155]: I see now, it propagates through the publisher/subscriber anyway and the extra propagation in fromSinkAndSource is not needed. 👍 .

File: project/StreamOperatorsIndexGenerator.scala [hunk 3]
```diff
@@ -83,7 +83,8 @@ object StreamOperatorsIndexGenerator extends AutoPlugin {
       "of",
       "join",
       "joinMat",
-      "fromFunction")),
+      "fromFunction",
+      "preMaterialize")),
```
[the remark this answers is not in the document]
Comment (johanandren) [reply project/StreamOperatorsIndexGenerator.scala:87]: Better add just a stub page with minimal docs in this PR than adding to skip list here.

**Review by johanandren — COMMENTED** [review]
Looks good but I think we should add a minimal doc page already here, doesn't need to have samples and such just the bare minimum short description and RS semantics block

**Review by johanandren — APPROVED** [review]

**Comment by hepin1989** [comment]
@patriknw ping~~

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**